tcmur_device.c

```diff
@@ -74,13 +61,28 @@ int __tcmu_reopen_dev(struct tcmu_device *dev, int retries)
 		 */
 		tcmu_dev_flush_ring(dev);
 
+	/*
+	 * To make sure all the in-flight IOs have been finished before
+	 * flushing the event_work. Or just after we flush the event_work
+	 * a new timedout IO callback could come and then it will fire a
+	 * new event work, which may access the device while we are closing
+	 * the device later.
+	 */
 	tcmu_dev_dbg(dev, "Waiting for outstanding commands to complete\n");
 	ret = aio_wait_for_empty_queue(rdev);
 	if (ret) {
 		pthread_mutex_lock(&rdev->rdev_lock);
 		goto done;
 	}
 
+	tcmur_flush_work(rdev->event_work);
+
+	pthread_mutex_lock(&rdev->rdev_lock);
+	if (rdev->flags & TCMUR_DEV_FLAG_IS_OPEN)
+		needs_close = true;
+	rdev->flags &= ~TCMUR_DEV_FLAG_IS_OPEN;
+	pthread_mutex_unlock(&rdev->rdev_lock);
```

Comment:
@lxbsz 
Can we close the handler under this lock, so that there are zero windows for any IO to happen?

Reply:
The lock isn't held across IO itself, so holding it for handler's `close` method would seem pointless.  The important thing is to ensure that nothing new can start executing in the interim and that appears to be achieved via flags.

Reply:
Yeah, agree with Ilya.